tsweb/debug.go

```diff
@@ -139,3 +148,161 @@ func gcHandler(w http.ResponseWriter, r *http.Request) {
+//      flagset.Int("max", 0, "maximum number of bars")
+//      flagset.String("s", "qux", "default name for new foos")
+//      go func() {
+//          for change := range flagc {
+//              // TODO: handle change, which will contain values for keys "max" and/or "s"
+//          }
+//      }()
+func (d *DebugHandler) FlagSet() (fs *flag.FlagSet, c chan map[string]interface{}) {
+	d.flagmu.Lock()
+	defer d.flagmu.Unlock()
+	if d.flagset == nil {
+		d.flagset = flag.NewFlagSet("debug", flag.ContinueOnError)
+		d.flagc = make(chan map[string]interface{})
+		d.Handle("flags", "Runtime flags", http.HandlerFunc(d.handleFlags))
+	}
+	return d.flagset, d.flagc
+}
+
+type copiedFlag struct {
+	Name  string
+	Value string
+	Usage string
+}
+
+func copyFlags(fs *flag.FlagSet) []copiedFlag {
+	var all []copiedFlag
+	fs.VisitAll(func(f *flag.Flag) {
+		all = append(all, copiedFlag{Name: f.Name, Value: f.Value.String(), Usage: f.Usage})
+	})
+	return all
+}
+
+func (d *DebugHandler) handleFlags(w http.ResponseWriter, r *http.Request) {
+	d.flagmu.Lock()
+	defer d.flagmu.Unlock()
+
+	var userError string
+	var modified string
+	if r.Method == http.MethodPost {
+		err := r.ParseForm()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		// Make a copy of existing values, in case we need to roll back.
+		all := copyFlags(d.flagset)
+		// Set inbound values.
+		changed := make(map[string][2]string)
+		rollback := false
+		for k, v := range r.PostForm {
+			if len(v) != 1 {
+				userError = fmt.Sprintf("multiple values for name %q: %q", k, v)
+				rollback = true
+				break
+			}
+			f := d.flagset.Lookup(k)
+			if f == nil {
+				userError = fmt.Sprintf("unknown name %q", k)
+				rollback = true
+				break
+			}
+			prev := f.Value.String()
+			new := strings.TrimSpace(v[0])
+			if prev == new {
+				continue
+			}
+			err := d.flagset.Set(k, new)
+			if err != nil {
+				userError = fmt.Sprintf("parsing value %q for name %q: %v", new, k, err)
+				rollback = true
+				break
+			}
+			changed[k] = [2]string{prev, new}
+		}
+		if rollback {
+			for _, f := range all {
+				d.flagset.Set(f.Name, f.Value)
+			}
+		} else {
+			// Generate description of modifications.
+			var names []string
+			for k := range changed {
+				names = append(names, k)
+			}
+			sort.Strings(names)
+			buf := new(strings.Builder)
+			for i, k := range names {
+				if i != 0 {
+					buf.WriteString("; ")
+				}
+				pn := changed[k]
+				fmt.Fprintf(buf, "%q: %v → %v", k, pn[0], pn[1])
+			}
+			modified = buf.String()
+			vals := make(map[string]interface{})
+			for _, k := range names {
+				vals[k] = d.flagset.Lookup(k).Value.(flag.Getter).Get()
+			}
+
+			d.flagc <- vals
+			// TODO: post modifications to Slack, along with attribution
+		}
+	}
+
+	dot := &struct {
+		Error    string
+		Modified string
+		Flags    []copiedFlag
+	}{
+		Error:    userError,
+		Modified: modified,
+		Flags:    copyFlags(d.flagset),
```

Comment:
What's the fidelity of the round-trip stringification and flag parsing? That is, we don't have confidence whether `flag.Value.String` will be parsed by `flag.Value.Set` to be identical to `flag.Value.` I think we should have a sanity check that `value.Set(value.String()).String() == value.String()` before proceeding here.

I'm concerned that trying to change flag X accidentally changes flag Y because of the lack of round-trip fidelity.

Reply:
A test like this also guarantees that the roll-back logic on L236 would succeed.

Reply:
The thought of a non-round-tripping flag.Value makes me shudder in horror. But sure, will add a check.